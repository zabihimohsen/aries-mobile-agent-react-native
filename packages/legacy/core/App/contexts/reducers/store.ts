import AsyncStorage from '@react-native-async-storage/async-storage'

import { LocalStorageKeys } from '../../constants'
import {
  Preferences as PreferencesState,
  Tours as ToursState,
  Onboarding as OnboardingState,
  Authentication as AuthenticationState,
  Lockout as LockoutState,
  LoginAttempt as LoginAttemptState,
  Migration as MigrationState,
  State,
} from '../../types/state'
import { generateRandomWalletName } from '../../utils/helpers'

enum OnboardingDispatchAction {
  ONBOARDING_UPDATED = 'onboarding/onboardingStateLoaded',
  DID_SEE_PREFACE = 'onboarding/didSeePreface',
  DID_COMPLETE_TUTORIAL = 'onboarding/didCompleteTutorial',
  DID_AGREE_TO_TERMS = 'onboarding/didAgreeToTerms',
  DID_CREATE_PIN = 'onboarding/didCreatePIN',
  DID_NAME_WALLET = 'onboarding/didNameWallet',
}

enum MigrationDispatchAction {
  DID_MIGRATE_TO_ASKAR = 'migration/didMigrateToAskar',
  MIGRATION_UPDATED = 'migration/migrationStateLoaded',
}

enum LockoutDispatchAction {
  LOCKOUT_UPDATED = 'lockout/lockoutUpdated',
}

enum LoginAttemptDispatchAction {
  ATTEMPT_UPDATED = 'loginAttempt/loginAttemptUpdated',
}

enum PreferencesDispatchAction {
  ENABLE_DEVELOPER_MODE = 'preferences/enableDeveloperMode',
  USE_BIOMETRY = 'preferences/useBiometry',
  PREFERENCES_UPDATED = 'preferences/preferencesStateLoaded',
  USE_VERIFIER_CAPABILITY = 'preferences/useVerifierCapability',
  USE_CONNECTION_INVITER_CAPABILITY = 'preferences/useConnectionInviterCapability',
  USE_DEV_VERIFIER_TEMPLATES = 'preferences/useDevVerifierTemplates',
  ENABLE_WALLET_NAMING = 'preferences/enableWalletNaming',
  UPDATE_WALLET_NAME = 'preferences/updateWalletName',
  ACCEPT_DEV_CREDENTIALS = 'preferences/acceptDevCredentials',
  USE_DATA_RETENTION = 'preferences/useDataRetention',
  PREVENT_AUTO_LOCK = 'preferences/preventAutoLock',
  UPDATE_ALTERNATE_CONTACT_NAMES = 'preferences/updateAlternateContactNames',
}

enum ToursDispatchAction {
  TOUR_DATA_UPDATED = 'tours/tourDataUpdated',
  UPDATE_SEEN_TOUR_PROMPT = 'tours/seenTourPrompt',
  ENABLE_TOURS = 'tours/enableTours',
  UPDATE_SEEN_HOME_TOUR = 'tours/seenHomeTour',
  UPDATE_SEEN_CREDENTIALS_TOUR = 'tours/seenCredentialsTour',
  UPDATE_SEEN_CREDENTIAL_OFFER_TOUR = 'tours/seenCredentialOfferTour',
  UPDATE_SEEN_PROOF_REQUEST_TOUR = 'tours/seenProofRequestTour',
}

enum AuthenticationDispatchAction {
  DID_AUTHENTICATE = 'authentication/didAuthenticate',
}

enum DeepLinkDispatchAction {
  ACTIVE_DEEP_LINK = 'deepLink/activeDeepLink',
}

export type DispatchAction =
  | OnboardingDispatchAction
  | LoginAttemptDispatchAction
  | LockoutDispatchAction
  | PreferencesDispatchAction
  | ToursDispatchAction
  | AuthenticationDispatchAction
  | DeepLinkDispatchAction
  | MigrationDispatchAction

export const DispatchAction = {
  ...OnboardingDispatchAction,
  ...LoginAttemptDispatchAction,
  ...LockoutDispatchAction,
  ...PreferencesDispatchAction,
  ...ToursDispatchAction,
  ...AuthenticationDispatchAction,
  ...DeepLinkDispatchAction,
  ...MigrationDispatchAction,
}

export interface ReducerAction<R> {
  type: R
  payload?: Array<any>
}

export const reducer = <S extends State>(state: S, action: ReducerAction<DispatchAction>): S => {
  switch (action.type) {
    case PreferencesDispatchAction.ENABLE_DEVELOPER_MODE: {
      const choice = (action?.payload ?? []).pop() ?? false
      const preferences = { ...state.preferences, developerModeEnabled: choice }

      AsyncStorage.setItem(LocalStorageKeys.Preferences, JSON.stringify(preferences))

      return {
        ...state,
        preferences,
      }
    }
    case PreferencesDispatchAction.USE_BIOMETRY: {
      const choice = (action?.payload ?? []).pop() ?? false
      const preferences = {
        ...state.preferences,
        useBiometry: choice,
      }
      const onboarding = {
        ...state.onboarding,
        didConsiderBiometry: true,
      }
      const newState = {
        ...state,
        onboarding,
        preferences,
      }

      AsyncStorage.setItem(LocalStorageKeys.Onboarding, JSON.stringify(onboarding))
      AsyncStorage.setItem(LocalStorageKeys.Preferences, JSON.stringify(preferences))

      return newState
    }
    case PreferencesDispatchAction.USE_VERIFIER_CAPABILITY: {
      const choice = (action?.payload ?? []).pop() ?? false
      const preferences = {
        ...state.preferences,
        useVerifierCapability: choice,
      }
      const newState = {
        ...state,
        preferences,
      }

      AsyncStorage.setItem(LocalStorageKeys.Preferences, JSON.stringify(preferences))

      return newState
    }
    case PreferencesDispatchAction.USE_CONNECTION_INVITER_CAPABILITY: {
      const choice = (action?.payload ?? []).pop() ?? false
      const preferences = {
        ...state.preferences,
        useConnectionInviterCapability: choice,
      }
      const newState = {
        ...state,
        preferences,
      }

      AsyncStorage.setItem(LocalStorageKeys.Preferences, JSON.stringify(preferences))

      return newState
    }
    case PreferencesDispatchAction.USE_DEV_VERIFIER_TEMPLATES: {
      const choice = (action?.payload ?? []).pop() ?? false
      const preferences = {
        ...state.preferences,
        useDevVerifierTemplates: choice,
      }
      const newState = {
        ...state,
        preferences,
      }

      AsyncStorage.setItem(LocalStorageKeys.Preferences, JSON.stringify(preferences))

      return newState
    }
    case PreferencesDispatchAction.ACCEPT_DEV_CREDENTIALS: {
      const choice = (action?.payload ?? []).pop() ?? false
      const preferences = {
        ...state.preferences,
        acceptDevCredentials: choice,
      }
      const newState = {
        ...state,
        preferences,
      }

      AsyncStorage.setItem(LocalStorageKeys.Preferences, JSON.stringify(preferences))

      return newState
    }
    case PreferencesDispatchAction.USE_DATA_RETENTION: {
      const choice = (action?.payload ?? []).pop() ?? false
      const preferences = {
        ...state.preferences,
        useDataRetention: choice,
      }
      const newState = {
        ...state,
        preferences,
      }

      AsyncStorage.setItem(LocalStorageKeys.Preferences, JSON.stringify(preferences))

      return newState
    }
    case PreferencesDispatchAction.PREFERENCES_UPDATED: {
      const preferences: PreferencesState = (action?.payload || []).pop()
      // For older wallets that haven't explicitly named their wallet yet
      if (!preferences.walletName) {
        preferences.walletName = generateRandomWalletName()
      }
      // For older wallets initialized before data retention option was created
      if (preferences.useDataRetention === undefined) {
        preferences.useDataRetention = true
      }

      if (!preferences.alternateContactNames) {
        preferences.alternateContactNames = {}
      }

      return {
        ...state,
        preferences,
      }
    }
    case PreferencesDispatchAction.ENABLE_WALLET_NAMING: {
      const choice = (action?.payload ?? []).pop() ?? false
      const preferences = { ...state.preferences, enableWalletNaming: choice }

      AsyncStorage.setItem(LocalStorageKeys.Preferences, JSON.stringify(preferences))

      return {
        ...state,
        preferences,
      }
    }
    case PreferencesDispatchAction.UPDATE_WALLET_NAME: {
      // We should never see 'My Wallet - 123', that's just there to let us know something went wrong while saving the wallet name
      const name = (action?.payload ?? []).pop() ?? 'My Wallet - 123'
      const preferences = {
        ...state.preferences,
        walletName: name,
      }
      const newState = {
        ...state,
        preferences,
      }

      AsyncStorage.setItem(LocalStorageKeys.Preferences, JSON.stringify(preferences))

      return newState
    }
    case PreferencesDispatchAction.PREVENT_AUTO_LOCK: {
      const choice = (action?.payload ?? []).pop() ?? false
      const preferences = { ...state.preferences, preventAutoLock: choice }

      AsyncStorage.setItem(LocalStorageKeys.Preferences, JSON.stringify(preferences))

      return {
        ...state,
        preferences,
      }
    }
    case PreferencesDispatchAction.UPDATE_ALTERNATE_CONTACT_NAMES: {
      const idNamePair = (action?.payload ?? []).pop() ?? {}
      const preferences = {
        ...state.preferences,
        alternateContactNames: {
          ...state.preferences.alternateContactNames,
          ...idNamePair,
        },
      }
      const newState = {
        ...state,
        preferences,
      }

      AsyncStorage.setItem(LocalStorageKeys.Preferences, JSON.stringify(preferences))

      return newState
    }
    case ToursDispatchAction.UPDATE_SEEN_TOUR_PROMPT: {
      const seenToursPrompt: ToursState = (action?.payload ?? []).pop() ?? false
      const tours = {
        ...state.tours,
        seenToursPrompt,
      }
      const newState = {
        ...state,
        tours,
      }

      return newState
    }
    case ToursDispatchAction.TOUR_DATA_UPDATED: {
      const tourData: ToursState = (action?.payload ?? []).pop() ?? {}
      const tours = {
        ...state.tours,
        ...tourData,
      }
      const newState = {
        ...state,
        tours,
      }

      return newState
    }
    case ToursDispatchAction.ENABLE_TOURS: {
      const enableTours = (action?.payload ?? []).pop() ?? false
      const tours = {
        ...state.tours,
      }
      if (enableTours) {
        tours.enableTours = enableTours
        tours.seenHomeTour = false
        tours.seenCredentialsTour = false
        tours.seenCredentialOfferTour = false
        tours.seenProofRequestTour = false
      } else {
        tours.enableTours = enableTours
      }
      const newState = {
        ...state,
        tours,
      }

      AsyncStorage.setItem(LocalStorageKeys.Tours, JSON.stringify(tours))

      return newState
    }
    case ToursDispatchAction.UPDATE_SEEN_HOME_TOUR: {
      const seenHomeTour = (action?.payload ?? []).pop() ?? false
      const tours = {
        ...state.tours,
        seenHomeTour,
      }

      if (seenHomeTour && tours.seenCredentialsTour && tours.seenCredentialOfferTour && tours.seenProofRequestTour) {
        tours.enableTours = false
      }

      const newState = {
        ...state,
        tours,
      }

      AsyncStorage.setItem(LocalStorageKeys.Tours, JSON.stringify(tours))

      return newState
    }
    case ToursDispatchAction.UPDATE_SEEN_CREDENTIALS_TOUR: {
      const seenCredentialsTour = (action?.payload ?? []).pop() ?? false
      const tours = {
        ...state.tours,
        seenCredentialsTour,
      }

      if (seenCredentialsTour && tours.seenHomeTour && tours.seenCredentialOfferTour && tours.seenProofRequestTour) {
        tours.enableTours = false
      }

      const newState = {
        ...state,
        tours,
      }

      AsyncStorage.setItem(LocalStorageKeys.Tours, JSON.stringify(tours))

      return newState
    }
    case ToursDispatchAction.UPDATE_SEEN_CREDENTIAL_OFFER_TOUR: {
      const seenCredentialOfferTour = (action?.payload ?? []).pop() ?? false
      const tours = {
        ...state.tours,
        seenCredentialOfferTour,
      }

      if (seenCredentialOfferTour && tours.seenHomeTour && tours.seenCredentialsTour && tours.seenProofRequestTour) {
        tours.enableTours = false
      }

      const newState = {
        ...state,
        tours,
      }

      AsyncStorage.setItem(LocalStorageKeys.Tours, JSON.stringify(tours))

      return newState
    }
    case ToursDispatchAction.UPDATE_SEEN_PROOF_REQUEST_TOUR: {
      const seenProofRequestTour = (action?.payload ?? []).pop() ?? false
      const tours = {
        ...state.tours,
        seenProofRequestTour,
      }

      if (seenProofRequestTour && tours.seenHomeTour && tours.seenCredentialOfferTour && tours.seenCredentialsTour) {
        tours.enableTours = false
      }

      const newState = {
        ...state,
        tours,
      }

      AsyncStorage.setItem(LocalStorageKeys.Tours, JSON.stringify(tours))

      return newState
    }

    case LoginAttemptDispatchAction.ATTEMPT_UPDATED: {
      const loginAttempt: LoginAttemptState = (action?.payload || []).pop()
      const newState = {
        ...state,
        loginAttempt,
      }
      AsyncStorage.setItem(LocalStorageKeys.LoginAttempts, JSON.stringify(newState.loginAttempt))
      return newState
    }
    case LockoutDispatchAction.LOCKOUT_UPDATED: {
      const lockout: LockoutState = (action?.payload || []).pop()
      return {
        ...state,
        lockout,
      }
    }
    case OnboardingDispatchAction.ONBOARDING_UPDATED: {
      const onboarding: OnboardingState = (action?.payload || []).pop()
      return {
        ...state,
        onboarding,
      }
    }
    case OnboardingDispatchAction.DID_SEE_PREFACE: {
      const onboarding = {
        ...state.onboarding,
        didSeePreface: true,
      }
      const newState = {
        ...state,
        onboarding,
      }
      AsyncStorage.setItem(LocalStorageKeys.Onboarding, JSON.stringify(newState.onboarding))
      return newState
    }
    case OnboardingDispatchAction.DID_COMPLETE_TUTORIAL: {
      const onboarding = {
        ...state.onboarding,
        didCompleteTutorial: true,
      }
      const newState = {
        ...state,
        onboarding,
      }
      AsyncStorage.setItem(LocalStorageKeys.Onboarding, JSON.stringify(newState.onboarding))
      return newState
    }
    case OnboardingDispatchAction.DID_AGREE_TO_TERMS: {
      const onboarding: OnboardingState = {
        ...state.onboarding,
        didAgreeToTerms: true,
      }
      const newState = {
        ...state,
        onboarding,
      }
      AsyncStorage.setItem(LocalStorageKeys.Onboarding, JSON.stringify(newState.onboarding))
      return newState
    }
    case OnboardingDispatchAction.DID_CREATE_PIN: {
      const onboarding: OnboardingState = {
        ...state.onboarding,
        didCreatePIN: true,
      }
      // If the pin is created with this version (that includes Askar), we
      // we can assume that a wallet using Indy SDK was never created. This
      // allows us to skip the migration step. For wallets initialized using
      // the indy sdk this won't be called, so the migration will be performed
      const migration: MigrationState = {
        ...state.migration,
        didMigrateToAskar: true,
      }
      const newState = {
        ...state,
        onboarding,
        migration,
      }
      AsyncStorage.setItem(LocalStorageKeys.Onboarding, JSON.stringify(newState.onboarding))
      AsyncStorage.setItem(LocalStorageKeys.Migration, JSON.stringify(newState.migration))
      return newState
    }
    case OnboardingDispatchAction.DID_NAME_WALLET: {
      const onboarding = {
        ...state.onboarding,
        didNameWallet: true,
      }
      const newState = {
        ...state,
        onboarding,
      }

      AsyncStorage.setItem(LocalStorageKeys.Onboarding, JSON.stringify(onboarding))

      return newState
    }
    case MigrationDispatchAction.DID_MIGRATE_TO_ASKAR: {
      const migration: MigrationState = {
        ...state.migration,
        didMigrateToAskar: true,
      }
      const newState = {
        ...state,
        migration,
      }
      AsyncStorage.setItem(LocalStorageKeys.Migration, JSON.stringify(newState.migration))
      return newState
    }
    case MigrationDispatchAction.MIGRATION_UPDATED: {
      const migration: MigrationState = (action?.payload || []).pop()
      return {
        ...state,
        migration,
      }
    }
    case AuthenticationDispatchAction.DID_AUTHENTICATE: {
      const value: AuthenticationState = (action?.payload || []).pop()
      const payload = value ?? { didAuthenticate: true }
      const newState = {
        ...state,
        ...{ authentication: payload },
      }
      return newState
    }
    case DeepLinkDispatchAction.ACTIVE_DEEP_LINK: {
      const value = (action?.payload || []).pop()
      return {
        ...state,
        ...{ deepLink: { activeDeepLink: value } },
      }
    }
    default:
      return state
  }
}

export default reducer
