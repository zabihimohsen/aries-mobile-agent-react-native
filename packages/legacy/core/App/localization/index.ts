import AsyncStorage from '@react-native-async-storage/async-storage'
// eslint-disable-next-line import/no-named-as-default
import i18n from 'i18next'
import { initReactI18next } from 'react-i18next'
import * as RNLocalize from 'react-native-localize'

import en from './en'
import fr from './fr'
import ptBr from './pt-br'

export type Translation = typeof en

export type TranslationResources = {
  [key: string]: any
}

export const translationResources: TranslationResources = {
  en: {
    translation: en,
  },
  fr: {
    translation: fr,
  },
  'pt-BR': {
    translation: ptBr,
  },
}

export enum Locales {
  en = 'en',
  fr = 'fr',
  ptBr = 'pt-BR',
}

const currentLanguage = i18n.language

//** Store language into the AsyncStorage  */
const storeLanguage = async (id: string) => {
  await AsyncStorage.setItem('language', id)
}

const initLanguages = (resources: TranslationResources, fallbackLng: Locales = Locales.en) => {
  const availableLanguages = Object.keys(resources)
  const bestLanguageMatch = RNLocalize.findBestAvailableLanguage(availableLanguages)
  let translationToUse = fallbackLng as string

  if (bestLanguageMatch && availableLanguages.includes(bestLanguageMatch.languageTag)) {
    translationToUse = bestLanguageMatch.languageTag
  }

  i18n.use(initReactI18next).init({
    debug: true,
    lng: translationToUse,
    fallbackLng,
    resources,
  })
}

//** Fetch user preference language from the AsyncStorage and set if require  */
const initStoredLanguage = async () => {
  const langId = await AsyncStorage.getItem('language')
  if (langId && langId !== currentLanguage) {
    await i18n.changeLanguage(langId)
  }
}

export { i18n, initStoredLanguage, initLanguages, storeLanguage, currentLanguage }
