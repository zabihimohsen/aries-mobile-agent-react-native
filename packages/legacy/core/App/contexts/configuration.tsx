import { IndyVdrPoolConfig } from '@aries-framework/indy-vdr'
import { OCABundleResolverType } from '@hyperledger/aries-oca/build/legacy'
import { createContext, ReducerAction, useContext } from 'react'

import { ProofRequestTemplate } from '../../verifier'
import { EmptyListProps } from '../components/misc/EmptyList'
import { RecordProps } from '../components/record/Record'
import { Locales } from '../localization'
import OnboardingPages from '../screens/OnboardingPages'
import { ScanProps } from '../screens/Scan'
import { PINSecurityParams } from '../types/security'
import { SettingSection } from '../types/settings'

interface NotificationConfiguration {
  component: React.FC
  onCloseAction: (dispatch?: React.Dispatch<ReducerAction<any>>) => void
  title: string
  description: string
  buttonTitle: string
  pageTitle: string
}

export interface ConfigurationContext {
  pages: typeof OnboardingPages
  splash: React.FC
  terms: React.FC
  preface: React.FC
  homeHeaderView: React.FC
  homeFooterView: React.FC
  credentialListHeaderRight: React.FC
  credentialListOptions: React.FC
  credentialEmptyList: React.FC<EmptyListProps>
  developer: React.FC
  OCABundleResolver: OCABundleResolverType
  proofTemplateBaseUrl?: string
  scan: React.FC<ScanProps>
  useBiometry: React.FC
  record: React.FC<RecordProps>
  PINSecurity: PINSecurityParams
  indyLedgers: IndyVdrPoolConfig[]
  settings: SettingSection[]
  customNotification: NotificationConfiguration
  useCustomNotifications: () => { total: number; notifications: any }
  supportedLanguages: Locales[]
  connectionTimerDelay?: number
  autoRedirectConnectionToHome?: boolean
  proofRequestTemplates?: (useDevTemplates: boolean) => Array<ProofRequestTemplate>
  enableTours?: boolean
  enableImplicitInvitations?: boolean
  enableReuseConnections?: boolean
  showPreface?: boolean
}

export const ConfigurationContext = createContext<ConfigurationContext>(null as unknown as ConfigurationContext)

export const ConfigurationProvider = ConfigurationContext.Provider

export const useConfiguration = () => useContext(ConfigurationContext)
