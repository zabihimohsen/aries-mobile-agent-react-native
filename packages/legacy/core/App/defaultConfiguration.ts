import { IOverlayBundleData } from '@hyperledger/aries-oca'
import { BrandingOverlayType, DefaultOCABundleResolver } from '@hyperledger/aries-oca/build/legacy'

import defaultIndyLedgers from '../configs/ledgers/indy'
import { useProofRequestTemplates } from '../verifier'

import * as bundle from './assets/oca-bundles.json'
import EmptyList from './components/misc/EmptyList'
import Record from './components/record/Record'
import HomeFooterView from './components/views/HomeFooterView'
import HomeHeaderView from './components/views/HomeHeaderView'
import { PINRules } from './constants'
import { ConfigurationContext } from './contexts/configuration'
import { useNotifications } from './hooks/notifications'
import { Locales, translationResources } from './localization'
import Developer from './screens/Developer'
import OnboardingPages from './screens/OnboardingPages'
import Preface from './screens/Preface'
import Scan from './screens/Scan'
import Splash from './screens/Splash'
import Terms from './screens/Terms'
import UseBiometry from './screens/UseBiometry'

export const defaultConfiguration: ConfigurationContext = {
  pages: OnboardingPages,
  splash: Splash,
  terms: Terms,
  preface: Preface,
  developer: Developer,
  homeHeaderView: HomeHeaderView,
  homeFooterView: HomeFooterView,
  credentialListHeaderRight: () => null,
  credentialListOptions: () => null,
  credentialEmptyList: EmptyList,
  OCABundleResolver: new DefaultOCABundleResolver(bundle as unknown as Record<string, IOverlayBundleData>, {
    brandingOverlayType: BrandingOverlayType.Branding10,
  }),
  scan: Scan,
  useBiometry: UseBiometry,
  record: Record,
  PINSecurity: { rules: PINRules, displayHelper: false },
  indyLedgers: defaultIndyLedgers,
  settings: [],
  customNotification: {
    component: () => null,
    onCloseAction: () => null,
    title: '',
    description: '',
    buttonTitle: '',
    pageTitle: '',
  },
  useCustomNotifications: useNotifications,
  proofRequestTemplates: useProofRequestTemplates,
  enableTours: false,
  supportedLanguages: Object.keys(translationResources) as Locales[],
  showPreface: false,
}
