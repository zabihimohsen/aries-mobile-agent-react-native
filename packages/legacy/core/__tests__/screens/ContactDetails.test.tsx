import { useConnectionById } from '@aries-framework/react-hooks'
import { useNavigation } from '@react-navigation/core'
import { StackNavigationProp } from '@react-navigation/stack'
import { act, fireEvent, render } from '@testing-library/react-native'
import fs from 'fs'
import path from 'path'
import React from 'react'

import { StoreContext } from '../../App'
import ContactDetails from '../../App/screens/ContactDetails'
import { ContactStackParams, Screens } from '../../App/types/navigators'
import { testIdWithKey } from '../../App/utils/testable'
import { testDefaultState } from '../contexts/store'

jest.mock('@react-navigation/core', () => {
  return require('../../__mocks__/custom/@react-navigation/core')
})
jest.mock('@react-navigation/native', () => {
  return require('../../__mocks__/custom/@react-navigation/native')
})
jest.mock('react-native-fs', () => ({}))
jest.mock('@hyperledger/anoncreds-react-native', () => ({}))
jest.mock('@hyperledger/aries-askar-react-native', () => ({}))
jest.mock('@hyperledger/indy-vdr-react-native', () => ({}))
jest.mock('react-native-permissions', () => require('react-native-permissions/mock'))
jest.mock('react-native-device-info', () => {
  return {
    getVersion: () => 1,
    getBuildNumber: () => 1,
  }
})

const connectionPath = path.join(__dirname, '../fixtures/connection-v1.json')
const connection = JSON.parse(fs.readFileSync(connectionPath, 'utf8'))
connection.createdAt = new Date('20230303')

describe('Contact Details Screen', () => {
  beforeEach(() => {
    jest.clearAllMocks()
    // @ts-ignore-next-line
    useConnectionById.mockReturnValue(connection)
  })

  test('Renders correctly', async () => {
    const tree = render(
      <StoreContext.Provider
        value={[
          testDefaultState,
          () => {
            return
          },
        ]}
      >
        <ContactDetails navigation={useNavigation()} route={{ params: { connectionId: connection.id } } as any} />
      </StoreContext.Provider>
    )
    expect(tree).toMatchSnapshot()
  })

  test('Renders correctly with alt contact name', async () => {
    const customState = {
      ...testDefaultState,
      preferences: {
        ...testDefaultState.preferences,
        alternateContactNames: {
          [connection.id]: 'Alt Name',
        },
      },
    }

    const tree = render(
      <StoreContext.Provider
        value={[
          customState,
          () => {
            return
          },
        ]}
      >
        <ContactDetails navigation={useNavigation()} route={{ params: { connectionId: connection.id } } as any} />
      </StoreContext.Provider>
    )
    expect(tree).toMatchSnapshot()
  })

  test('Edit contact name button navigates', async () => {
    const navigation = useNavigation<StackNavigationProp<ContactStackParams, Screens.ContactDetails>>()

    const { getByTestId } = render(
      <StoreContext.Provider
        value={[
          testDefaultState,
          () => {
            return
          },
        ]}
      >
        <ContactDetails navigation={navigation} route={{ params: { connectionId: connection.id } } as any} />
      </StoreContext.Provider>
    )
    const editButton = getByTestId(testIdWithKey('RenameContact'))
    await act(async () => {
      fireEvent(editButton, 'press')
      expect(navigation.navigate).toHaveBeenCalledTimes(1)
    })
  })
})
