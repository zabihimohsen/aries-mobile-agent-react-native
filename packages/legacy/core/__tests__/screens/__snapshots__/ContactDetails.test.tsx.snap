// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Contact Details Screen Renders correctly 1`] = `
<RNCSafeAreaView
  edges={
    Array [
      "bottom",
      "left",
      "right",
    ]
  }
  style={
    Object {
      "flexGrow": 1,
    }
  }
>
  <View
    style={
      Object {
        "backgroundColor": "#313132",
        "padding": 20,
      }
    }
  >
    <Text
      style={
        Object {
          "color": "#FFFFFF",
          "fontSize": 26,
          "fontWeight": "bold",
        }
      }
    >
      BestBC College
    </Text>
    <Text
      style={
        Object {
          "color": "#FFFFFF",
          "fontSize": 18,
          "fontWeight": "normal",
          "marginTop": 20,
        }
      }
    >
      ContactDetails.DateOfConnection
    </Text>
  </View>
  <View
    accessibilityLabel="Screens.RenameContact"
    accessibilityRole="button"
    accessibilityState={
      Object {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      Object {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "backgroundColor": "#313132",
        "marginTop": 10,
        "opacity": 1,
        "padding": 20,
      }
    }
    testID="com.ariesbifold:id/RenameContact"
  >
    <Text
      style={
        Object {
          "color": "#FFFFFF",
          "fontSize": 18,
          "fontWeight": "normal",
        }
      }
    >
      Screens.RenameContact
    </Text>
  </View>
  <View
    accessibilityLabel="ContactDetails.RemoveContact"
    accessibilityRole="button"
    accessibilityState={
      Object {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      Object {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "backgroundColor": "#313132",
        "marginTop": 10,
        "opacity": 1,
        "padding": 20,
      }
    }
    testID="com.ariesbifold:id/RemoveFromWallet"
  >
    <Text
      style={
        Object {
          "color": "#D8292F",
          "fontSize": 18,
          "fontWeight": "normal",
        }
      }
    >
      ContactDetails.RemoveContact
    </Text>
  </View>
  <Modal
    animationType="slide"
    hardwareAccelerated={false}
    transparent={true}
    visible={false}
  />
  <Modal
    animationType="slide"
    hardwareAccelerated={false}
    transparent={true}
    visible={false}
  />
</RNCSafeAreaView>
`;

exports[`Contact Details Screen Renders correctly with alt contact name 1`] = `
<RNCSafeAreaView
  edges={
    Array [
      "bottom",
      "left",
      "right",
    ]
  }
  style={
    Object {
      "flexGrow": 1,
    }
  }
>
  <View
    style={
      Object {
        "backgroundColor": "#313132",
        "padding": 20,
      }
    }
  >
    <Text
      style={
        Object {
          "color": "#FFFFFF",
          "fontSize": 26,
          "fontWeight": "bold",
        }
      }
    >
      Alt Name
    </Text>
    <Text
      style={
        Object {
          "color": "#FFFFFF",
          "fontSize": 18,
          "fontWeight": "normal",
          "marginTop": 20,
        }
      }
    >
      ContactDetails.DateOfConnection
    </Text>
  </View>
  <View
    accessibilityLabel="Screens.RenameContact"
    accessibilityRole="button"
    accessibilityState={
      Object {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      Object {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "backgroundColor": "#313132",
        "marginTop": 10,
        "opacity": 1,
        "padding": 20,
      }
    }
    testID="com.ariesbifold:id/RenameContact"
  >
    <Text
      style={
        Object {
          "color": "#FFFFFF",
          "fontSize": 18,
          "fontWeight": "normal",
        }
      }
    >
      Screens.RenameContact
    </Text>
  </View>
  <View
    accessibilityLabel="ContactDetails.RemoveContact"
    accessibilityRole="button"
    accessibilityState={
      Object {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      Object {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "backgroundColor": "#313132",
        "marginTop": 10,
        "opacity": 1,
        "padding": 20,
      }
    }
    testID="com.ariesbifold:id/RemoveFromWallet"
  >
    <Text
      style={
        Object {
          "color": "#D8292F",
          "fontSize": 18,
          "fontWeight": "normal",
        }
      }
    >
      ContactDetails.RemoveContact
    </Text>
  </View>
  <Modal
    animationType="slide"
    hardwareAccelerated={false}
    transparent={true}
    visible={false}
  />
  <Modal
    animationType="slide"
    hardwareAccelerated={false}
    transparent={true}
    visible={false}
  />
</RNCSafeAreaView>
`;
